### PR TITLE
Keep host port socket open for kubenet

### LINKED ISSUE #37087: host port is not hold by kubelet when using kubenet
I deploy an environment with kubenet according to this guide:
http://kubernetes.io/docs/admin/network-plugins/

Then I create a deployment with host port mapping:
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nginx
spec:
  replicas: 2
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
          hostPort: 80

Then I find that the 80 port in node is not hold by kubelet progress.

**Kubernetes version** (use `kubectl version`):
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-alpha.0.182+13cc43af826a5d", GitCommit:"13cc43af826a5d2769543e91d13c93dd68c6f809", GitTreeState:"clean", BuildDate:"2016-11-08T04:42:35Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-alpha.0.477+d5014c3f216204-dirty", GitCommit:"d5014c3f216204c2a8464a5bdf543f408eee6f71", GitTreeState:"dirty", BuildDate:"2016-11-14T11:01:41Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Cloud provider or hardware configuration**: ubuntu local machine
- **OS** (e.g. from /etc/os-release):  
NAME="Ubuntu"
VERSION="16.04.1 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.1 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
- **Kernel** (e.g. `uname -a`):
Linux kubenet-01 4.4.0-47-generic #68-Ubuntu SMP Wed Oct 26 19:39:52 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**What happened**:
The host port is not token by kubelet.

**What you expected to happen**:
The host port should be token by kubelet.

### PR DESCRIPTION
**Pull request by CaoShuFeng**
fixes #37087

**Release note**:

```release-note
NONE
```

When cni is set to kubenet, kubelet should hold the host port socket,
so that other application in this node could not listen/bind this port
any more. However, the sockets are closed accidentally, because
kubelet forget to reconcile the protocol format before comparing.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37088)

@kubernetes/sig-network

### CODE REVIEW

**Comment by dims** [comment]
@k8s-bot ok to test

**Comment by CaoShuFeng** [comment]
Sorry, I do a wrong rebase action.
So, many unrelated changes show up in this pr.
I will try to fix it.

**Comment by CaoShuFeng** [comment]
@k8s-bot verify test this

**Comment by CaoShuFeng** [comment]
@k8s-bot verify test this

**Comment by feiskyer** [comment]
@k8s-bot verify test this

File: pkg/kubelet/network/hostport/hostport_test.go [hunk 1]
```diff
@@ -185,6 +186,10 @@ func TestOpenPodHostports(t *testing.T) {
 		})
 	}
 
+	// Already running pod's host port
+	hp := hostport{tests[1].pod.Spec.Containers[0].Ports[0].HostPort, strings.ToLower(string(tests[1].pod.Spec.Containers[0].Ports[0].Protocol))}
+	h.hostPortMap[hp] = &fakeSocket{tests[1].pod.Spec.Containers[0].Ports[0].HostPort, strings.ToLower(string(tests[1].pod.Spec.Containers[0].Ports[0].Protocol)), false}
```
Comment (feiskyer) [inline pkg/kubelet/network/hostport/hostport_test.go]: nit: lines are too long, consider to multiple lines.

Comment (CaoShuFeng) [inline pkg/kubelet/network/hostport/hostport_test.go]: Done.

File: pkg/kubelet/network/hostport/hostport.go [hunk 2]
```diff
@@ -97,7 +97,8 @@ func (h *handler) openHostports(pod *api.Pod) error {
 				continue
 			}
 			hp := hostport{
-				port:     port.HostPort,
+				port: port.HostPort,
+				// When cleaning it up, we also check the protocal in lower format.
```
Comment (caseydavenport) [inline pkg/kubelet/network/hostport/hostport.go]: Can you please undo this change?

Comment (CaoShuFeng) [inline pkg/kubelet/network/hostport/hostport.go]: Done.

**Comment by freehan** [comment]
Thanks! LGTM

**Comment by freehan** [comment]
Applying 1.5 milestone since this is a bug fix.

**Comment by saad-ali** [comment]
> Applying 1.5 milestone since this is a bug fix.

Ack.

**Comment by feiskyer** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by CaoShuFeng** [comment]
Hi, the ci test says verify-gofmt.sh failed, but I have passed it in my local environment, can someone help me?
>I1119 13:15:47.721] !!! 'gofmt -s -w' needs to be run on the following files: 
>I1119 13:15:47.721] ./pkg/kubelet/network/hostport/hostport.go
>I1119 13:15:47.722] ./pkg/kubelet/network/hostport/hostport_test.go
>I1119 13:15:47.723] FAILED   hack/make-rules/../../hack/verify-gofmt.sh	9s

**Comment by feiskyer** [comment]
@CaoShuFeng Maybe try rebase and gofmt again?

**Comment by CaoShuFeng** [comment]
@feiskyer rebased and run "bash hack/verify-gofmt.sh" also ok.

**Comment by freehan** [comment]
@CaoShuFeng did you try a different go/gofmt version?

**Comment by CaoShuFeng** [comment]
I rebased it.
And run "make verify" "make test" successfully in my environment.
Hope this time it works.

**Comment by dims** [comment]
@k8s-bot test this

**Comment by dims** [comment]
@k8s-bot test this

**Comment by dims** [comment]
Dear @kubernetes/test-infra-maintainers Where can i look up why 10 of the jobs ended with "Error in queue"?

**Comment by krzyzacy** [comment]
the error message come from here: https://github.com/kubernetes/test-infra/blob/58bb3baabf5a67f0c86a8f73516d1d8c5d625303/prow/cmd/line/main.go#L285

maybe a jenkins request error, and seems it's after https://github.com/kubernetes/test-infra/pull/1215 is merged. @spxtr probably can explain better.

**Comment by dims** [comment]
@k8s-bot node e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot unit test this
@k8s-bot verify test this
@k8s-bot node e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot gce etcd3 e2e test this

**Comment by dims** [comment]
Requesting the minimum jobs needed for this to be picked up by SQ

**Comment by spxtr** [comment]
@dims feel free to open an issue in test-infra for this sort of thing in the future. We're looking into the "Error in queue" issue and hopefully it's fixed by now.

**Comment by dims** [comment]
Ack @spxtr will do!

**Comment by dims** [comment]
@k8s-bot test this [Multiple CI jobs failed]

**Comment by rmmh** [comment]
@k8s-bot test this issue kubernetes/test-infra#1253 (also, deleting spurious failure comments)

**Comment by dims** [comment]
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gke e2e test this

**Comment by saad-ali** [comment]
Does this need to be cherry-picked to 1.5? If so make sure to add `cherrypick-candidate` label

**Comment by saad-ali** [comment]
> Does this need to be cherry-picked to 1.5? If so make sure to add cherrypick-candidate label

Ping. CC @freehan
